teps/0041-move-image-entrypoint-lookup-to-the-taskrun-pod.md

```diff
@@ -0,0 +1,161 @@
+---
+status: proposed
+title: Move Image Entrypoint Lookup to the TaskRun Pod
+creation-date: '2021-01-11'
+last-updated: '2021-01-21'
+authors:
+- '@yaoxiaoqi'
+---
+
+# TEP-0041: Move Image Entrypoint Lookup to the TaskRun Pod
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+- [Proposal](#proposal)
+  - [Add Flags to EntryPoint Binary](#add-flags-to-entrypoint-binary)
+  - [Resolve Image in EntryPoint Binary](#resolve-image-in-entrypoint-binary)
+  - [Notes/Caveats](#notescaveats)
+    - [Entrypoint Cache](#entrypoint-cache)
+    - [Error Reporting Delay](#error-reporting-delay)
+    - [Inconsistency](#inconsistency)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes to move the image entrypoint lookup to entrypoint binary that
+runs in the TaskRun pod.
+
+## Motivation
+
+Currently, Tekton controller pod has access to all service accounts and will do
+entry point lookup using `go-containerregistry` for the images that don't have
+any command specified for them. This potentially might cause permission denied
+error when users are trying to pull private images from registry since the
+default service account and the service account that Tekton controller is using
+can't authenticate the requests on some of cloud platforms. The private image is
+actually running on the TaskRun pod so this will involve some inter-pod
+permission issue.
+
+Therefore, this TEP proposes to move the image entrypoint lookup to TaskRun pod.
+By doing this, we can use the service account that user provided for the TaskRun
+to query the entrypoint of the image that is running on this pod. This will also
+avoid the authentication problem when requests come from different pods on some
+cloud providers and the confusing retricted access to the image when command is not
+specified.
```

Comment:
can you elaborate on this authentication problem a bit?

Reply:
Sorry I didn't put it clearly in the TEP. I updated the section after I did some experiments on entrypoint lookup and reproduced the issue https://github.com/tektoncd/pipeline/issues/2316 using GKE cluster. When Workload Identity enabled, user has to use tekton pipelines controller service account to impersonate the corresponding GSA since the request comes from controller pod. But users usually don't know they should do this and only bind the service account for the target pod to GSA. Billy gave a better explanation here https://wlyn.ch/posts/gcp-tekton-workload-identity/
the experiments I did lists below:
| Cluster | Config  | Image fetching Status  | Note  |
|------|--------|----------|---------------|
| minikube | Running registry in the cluster | Successful | Must ensure the connection between controller pod and registry pod first |
| GKE | Workload identity disabled | Successful | GCE metadata server takes care of authentication  |
| GKE | Workload identity enabled | Failed | Only linked the target service account with GSA |
| GKE | Workload identity enabled | Successful | Only linked the tekton pipelines controller service account with GSA |